crates/telio-nurse/src/aggregator.rs

```diff
@@ -0,0 +1,808 @@
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            ConnectionData::Relay { state, reason } => {
+                let mut seq = serializer.serialize_tuple(RELAY_TUPLE_LEN)?;
+                seq.serialize_element(&(*state as u64))?;
+                seq.serialize_element(&(*reason as u64))?;
+                seq.end()
+            }
+            ConnectionData::Peer {
+                endpoints,
+                rx_bytes,
+                tx_bytes,
+            } => {
+                let initiator_repr = endpoints.initiator_ep as u16;
+                let responder_repr = endpoints.responder_ep as u16;
+                let strategy_id = (initiator_repr << ENDPOINT_BIT_FIELD_WIDTH) + responder_repr;
+                let mut seq = serializer.serialize_tuple(PEER_TUPLE_LEN)?;
+                seq.serialize_element(&strategy_id)?;
+                seq.serialize_element(&rx_bytes)?;
+                seq.serialize_element(&tx_bytes)?;
+                seq.end()
+            }
+        }
+    }
+}
+
+// Ready to send connection data for some period of time
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+pub struct ConnectionSegmentData {
+    node: PublicKey,
+    start: Instant, // to ensure the segments are unique
+    duration: Duration,
+    connection_data: ConnectionData,
+}
+
+impl ConnectionSegmentData {
+    pub fn new_relay(relay_event: &RelayEvent, target_timestamp: Instant) -> Self {
+        ConnectionSegmentData {
+            node: relay_event.public_key,
+            start: relay_event.timestamp,
+            duration: target_timestamp.duration_since(relay_event.timestamp),
+            connection_data: ConnectionData::Relay {
+                state: relay_event.peer_state.into(),
+                reason: relay_event.reason,
+            },
+        }
+    }
+
+    pub fn new_peer(
+        peer_event: &AnalyticsEvent,
+        endpoints: PeerEndpointTypes,
+        target_timestamp: Instant,
+        target_rx_bytes: u64,
+        target_tx_bytes: u64,
+    ) -> Self {
+        ConnectionSegmentData {
+            node: peer_event.public_key,
+            start: peer_event.timestamp.into(),
+            duration: target_timestamp.duration_since(peer_event.timestamp.into()),
+            connection_data: ConnectionData::Peer {
+                endpoints,
+                rx_bytes: target_rx_bytes - peer_event.rx_bytes,
+                tx_bytes: target_tx_bytes - peer_event.tx_bytes,
+            },
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct RelayEvent {
+    public_key: PublicKey,
+    peer_state: PeerState,
+    timestamp: Instant,
+    reason: RelayConnectionChangeReason,
+}
+
+struct AggregatorData {
+    current_relay_event: Option<RelayEvent>,
+    current_peer_events: HashMap<PublicKey, (AnalyticsEvent, PeerEndpointTypes)>,
+    unacknowledged_segments: Vec<ConnectionSegmentData>,
+}
```

Comment:
What will happen with this state when application is going to be shutdown?

Reply:
@Jauler what do you think about this one? It looks to me that while it should be handled in a bit different manner, it is kinda similar to the sleeping one 🤔

Reply:
There will be a responsibility of telio-nurse, to generate "meshnet disable" event, and collect those events from the aggregator.